### PR TITLE
feat(hive):  suport substr && order by with null values compatible with hive functions

### PR DESCRIPTION
**Pull request by sandflee**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

part: #8476

### CODE REVIEW

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 1]
```diff
@@ -928,6 +941,13 @@ impl<'a> TypeChecker<'a> {
             arg_types.push(arg_type);
         }
 
+        // rewrite substr('xx', 0, xx) -> substr('xx', 1, xx)
+        if self.ctx.get_settings().get_sql_dialect().unwrap() == Dialect::Hive
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check.rs:945]: new method like  `fn index_zero_literal_as_one(&self) -> bool`

Comment (sandflee) [reply src/query/sql/src/planner/semantic/type_check.rs:945]: add this method in Dialect?

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 2]
```diff
@@ -928,6 +940,18 @@ impl<'a> TypeChecker<'a> {
             arg_types.push(arg_type);
         }
 
+        // rewrite substr('xx', 0, xx) -> substr('xx', 1, xx)
+        if (func_name == "substr" || func_name == "substring")
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check.rs:944]: need a test case in  tests/logictest

Comment (sundy-li) [reply src/query/sql/src/planner/semantic/type_check.rs:944]: 02_0005_function_substring

File: tests/logictest/suites/query/02_function/02_0005_function_substring [hunk 3]
```diff
@@ -200,3 +200,17 @@ SELECT SUBSTR('12345', 2, 1);
 ----
 2
 
+
+statement query T
```
Comment (sundy-li) [inline tests/logictest/suites/query/02_function/02_0005_function_substring:204]:
```suggestion
statement ok
set sql_dialect = 'hive';

statement query T
SELECT SUBSTR('12345', 0, 1);

----
1
```

**Review by sundy-li — APPROVED** [review]

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**